File: ios/Mask.swift
public class Mask {
  public struct Options {
    let groupingSeparator: String
    let fractionSeparator: String
    let suffix: String?
    let prefix: String?
    let maximumIntegerDigits: Int?
    let maximumFractionalDigits: Int
    let minValue: Double?
    let maxValue: Double?
    var suffixLength: Int {
      get {
        suffix?.count ?? 0
      }
    }
    let needBeforeUnmasking: Bool
  }
  
  static let systemDecimalSeparator = NumberFormatter().decimalSeparator ?? "."
  static let jsDecimalSeparator = "."
  
  public static func apply(for text: String, withOptions options: Options) -> String {
    let rawText = Mask.range(
      text: options.needBeforeUnmasking ? Mask.unmask(text: text, withOptions: options) : text,
      minValue: options.minValue,
      maxValue: options.maxValue
    )
      .replace(substring: systemDecimalSeparator, withTemplate: options.fractionSeparator)
    
    if rawText.count == 0 {
      return ""
    }
    
    let fractionSeparators = rawText.components(separatedBy: options.fractionSeparator)
    
    let numberFormatter = NumberFormatter()
    
    let maximumIntegerDigits = options.maximumIntegerDigits ?? numberFormatter.maximumIntegerDigits
    
    numberFormatter.numberStyle = .decimal
    numberFormatter.maximumIntegerDigits = maximumIntegerDigits
    numberFormatter.groupingSeparator = options.groupingSeparator
    
    let integerPart = Decimal(string: truncate(fractionSeparators[0], toLength: maximumIntegerDigits)) ?? 0
    
    var result = numberFormatter.string(from: NSDecimalNumber(decimal: integerPart)) ?? ""
    
    if (options.maximumFractionalDigits > 0) {
      guard let fractionalPart = fractionSeparators[safe: 1] else {
        return result
          .with(prefix: options.prefix)
          .with(suffix: options.suffix)
      }
      
      result += options.fractionSeparator
      result += fractionalPart[safe: 0...options.maximumFractionalDigits]
    }
    
    return result
      .with(prefix: options.prefix)
      .with(suffix: options.suffix)
  }
  
  public static func unmask(text: String, withOptions options: Options) -> String {
    return text
      .replace(substring: options.groupingSeparator, withTemplate: "")
      .replace(substring: options.fractionSeparator, withTemplate: jsDecimalSeparator)
      .replace(prefix: options.prefix ?? "", withTemplate: "")
      .replace(suffix: options.suffix ?? "", withTemplate: "")
      .replace(pattern: "[^\\d\\\(jsDecimalSeparator)]", withTemplate: "")
  }
  
  private static func range(text: String, minValue: Double?, maxValue: Double?) -> String {
    var result = text == systemDecimalSeparator
    ? "0."
    : text
    
    if (minValue != nil) {
      guard let value = Double(result) else {
        return ""
      }
      
      if (minValue! > value) {
        result = minValue!.format()
      }
    }
    
    if (maxValue != nil) {
      guard let value = Double(result) else {
        return ""
      }
      
      if (maxValue! < value) {
        result = maxValue!.format()
      }
    }
    
    return result
      .replace(substring: ".", withTemplate: systemDecimalSeparator)
  }
  
  private static func truncate(_ string: String, toLength length: Int) -> String {
    return string.count > length
    ? String(string.prefix(length))
    : string
  }
}

File: android/src/main/java/com/alexzunik/reactnativemoneyinput/Mask.kt
package com.alexzunik.reactnativemoneyinput

import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import kotlin.math.floor

class Mask {
  class Options(
    val groupingSeparator: String,
    val fractionSeparator: String,
    val prefix: String,
    val suffix: String,
    val maximumIntegerDigits: Int?,
    val maximumFractionalDigits: Int,
    val minValue: Double?,
    val maxValue: Double?,
  ){}


  companion object {
    val systemDecimalSeparator = DecimalFormat().decimalFormatSymbols.decimalSeparator.toString()
    val jsDecimalSeparator = "."

    fun apply(forText: String, options: Options): String {
      val rawText = range(
        text =  unmask(forText, options),
        minValue = options.minValue,
        maxValue = options.maxValue
      )
        .replace(systemDecimalSeparator, options.fractionSeparator)

      if (rawText.isEmpty()) {
        return ""
      }

      val fractionSeparators = rawText.split(options.fractionSeparator)

      val decimalFormatSymbols = DecimalFormatSymbols()
      decimalFormatSymbols.groupingSeparator = options.groupingSeparator.toCharArray()[0]

      val formatter = DecimalFormat()

      val maximumIntegerDigits = options.maximumIntegerDigits ?: formatter.maximumIntegerDigits

      formatter.maximumIntegerDigits = maximumIntegerDigits
      formatter.decimalFormatSymbols = decimalFormatSymbols

      val integerPart = if (fractionSeparators[0].isEmpty()) 0
        else truncate(fractionSeparators[0], maximumIntegerDigits).toBigInteger()

      var result = formatter.format(integerPart)

      if (fractionSeparators.count() == 1) {
        return result
          .withPrefix(options.prefix)
          .withSuffix(options.suffix)
      }

      if (options.maximumFractionalDigits > 0) {
        result += options.fractionSeparator

        val fraction = fractionSeparators[1]

        result += if (fraction.count() > options.maximumFractionalDigits)
          fraction.slice(0 until options.maximumFractionalDigits)
        else
          fraction
      }


      return result
        .withPrefix(options.prefix)
        .withSuffix(options.suffix)
    }

    fun unmask(text: String, options: Options): String {
      return text
        .replace(options.fractionSeparator, jsDecimalSeparator)
        .removePrefix(options.prefix)
        .removeSuffix(options.suffix)
        .replace(Regex("[^\\d\\$jsDecimalSeparator]"), "")
    }

    private fun range(text: String, minValue: Double?, maxValue: Double?): String {
      var result = if (text == jsDecimalSeparator)
        "0."
      else
        text

      if (minValue != null) {
        val value = result.toDoubleOrNull() ?: return ""

        if (minValue > value) {
          result = minValue.format()
        }
      }

      if (maxValue != null) {
        val value = result.toDoubleOrNull() ?: return ""

        if (maxValue < value) {
          result = maxValue.format()
        }
      }

      return result
        .replace(".", systemDecimalSeparator)
    }

    private fun truncate(string: String, length: Int): String {
      return if (string.length > length) {
        string.substring(0, length)
      } else {
        string
      }
    }

    private fun Double.format(): String {
      val likeInteger = floor(this) == this
      return if (likeInteger)
          this.toString().replace(".0", "")
        else
          this.toString()
    }

    private fun String.withSuffix(suffix: String): String {
      if (this.isEmpty()) {
        return this
      }
      return this + suffix
    }

    private fun String.withPrefix(prefix: String): String {
      if (this.isEmpty()) {
        return this
      }
      return prefix + this
    }
  }
}
